micro_http/src/connection.rs

```diff
@@ -0,0 +1,675 @@
+                        .collect();
+                    request.body = Some(Body::new(placeholder));
+
+                    // If we read more bytes than we should have into the body of the request.
+                    if !self.body_vec.is_empty() {
+                        return Err(ConnectionError::StreamError);
+                    }
+
+                    self.state = ConnectionState::RequestReady;
+                }
+                ConnectionState::RequestReady => {
+                    // This request is ready to be passed for handling.
+                    // Update the state machine to expect a new request and push this request into
+                    // the `parsed_requests` queue.
+                    self.state = ConnectionState::WaitingForRequestLine;
+                    self.body_bytes_to_be_read = 0;
+                    self.parsed_requests
+                        .push_back(self.pending_request.take().unwrap());
+                }
+            };
+        }
+
+        Ok(())
+    }
+
+    /// Tries to write responses to the provided stream.
+    /// Meant to be used only with non-blocking streams and an `EPOLL` structure.
+    /// Should be called whenever an `EPOLLOUT` event is signaled.
+    pub fn try_write(&mut self) -> Result<(), ConnectionError> {
+        if self.response_buffer.is_none() {
+            if let Some(response) = self.response_queue.pop_front() {
+                let mut response_buffer_vec: Vec<u8> = Vec::new();
+                response
+                    .write_all(&mut response_buffer_vec)
+                    .map_err(|_| ConnectionError::StreamError)?;
+                self.response_buffer = Some(response_buffer_vec);
+            } else {
+                return Ok(());
+            }
+        }
+
+        let mut response_fully_written = false;
+
+        if let Some(response_buffer_vec) = self.response_buffer.as_mut() {
+            let bytes_to_be_written = response_buffer_vec.len();
+            match self.stream.write(response_buffer_vec.as_slice()) {
+                Ok(0) => {
+                    return Err(ConnectionError::ConnectionClosed);
+                }
+                Ok(bytes_written) => {
+                    if bytes_written != bytes_to_be_written {
+                        response_buffer_vec.drain(..bytes_written);
+                    } else {
+                        response_fully_written = true;
+                    }
+                }
+                Err(_) => {
+                    return Err(ConnectionError::StreamError);
+                }
+            }
+        }
+
+        if response_fully_written {
+            self.response_buffer.take();
+        }
+
+        Ok(())
+    }
+
+    /// Send a response back to the source of a request.
+    pub fn send_response(&mut self, response: Response) {
+        self.response_queue.push_back(response);
+    }
+
+    fn reset_buffer(&mut self, line_start_index: usize, bytes_read: usize) {
```

Comment:
I wouldn't call this `reset_buffer`. This name lacks a lot of info. Maybe `reset_buffer_to_substr` although it doesn't sound very good either. I would also rename the parameters:

```
fn reset_buffer_to_substr(&mut self, substr_start: usize, substr_end: usize)
```